### PR TITLE
feature: binary_operator_spaces - Revert change about => alignment and use option instead

### LINKED ISSUE #6716: New option to align arrows (binary_operator_spaces)
## Feature request

I would like the option for the behavior in v3.13.0 `binary_operator_spaces` `align_single_space_minimal` for arrays.
Ref: #6590

config:
```php
        'binary_operator_spaces' => [
            'default'   => 'single_space',
            'operators' => [
                '='  => 'align_single_space_minimal',
                '=>' => 'align_single_space_minimal',
                '||' => 'align_single_space_minimal',
                '.=' => 'align_single_space_minimal',
            ],
        ],
```

v3.13.0 result:
```php
$a = [
    'a'  => 'a',
    'aa' => 'aa',
    // bra bra bra
    'aaa' => 'aaa',
];
```

v3.13.1 result:
```php
$a = [
    'a'   => 'a',
    'aa'  => 'aa',
    // bra bra bra
    'aaa' => 'aaa',
];
```

### PR DESCRIPTION
**Pull request by VincentLanglet**
HI @julienfalque

The previous bugfix in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6590 is subject to debate about yes or no it was a bug.

The issue https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/4395 and https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/5546 consider it's a bug.
The issue https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6716 consider it's not.

I think the best solution is
- To revert #6590 for the option `align`, `align_single_space`, `align_single_space_minimal`
- To provide this behavior inside anothers options `align_by_scope`, `align_single_space_by_scope`, `align_single_space_minimal_by_scope`

This closes #6716

The only doubt I have is if `by_scope` is a good suffix or if there is one more meaningful...

### CODE REVIEW

File: tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php [hunk 1]
```diff
@@ -2312,9 +3009,25 @@ function foo () {
 }
 ',
             ],
+            [
+                '<?php
```
Comment (VincentLanglet) [inline tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php:3013]: Test added for https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6716#issuecomment-1364486054

File: src/Fixer/Operator/BinaryOperatorSpacesFixer.php [hunk 2]
```diff
@@ -656,6 +733,10 @@ private function injectAlignmentPlaceholdersForArrow(Tokens $tokens, int $startA
             // no need to analyze for `isBinaryOperator` (always true), nor if part of declare statement (not valid PHP)
             // there is also no need to analyse the second arrow of a line
             if ($token->isGivenKind(T_DOUBLE_ARROW) && $newLineFoundSinceLastPlaceholder) {
+                if ($yieldFoundSinceLastPlaceholder) {
```
Comment (VincentLanglet) [inline src/Fixer/Operator/BinaryOperatorSpacesFixer.php:736]: To handle https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6716#issuecomment-1364486054

When a yield was previously use, it's another scope/level.

File: src/Fixer/Operator/BinaryOperatorSpacesFixer.php [hunk 3]
```diff
@@ -774,7 +859,11 @@ private function replacePlaceholders(Tokens $tokens, string $alignStrategy, stri
             foreach ($lines as $index => $line) {
                 if (substr_count($line, $placeholder) > 0) {
                     $groups[$groupIndex][] = $index;
-                } elseif ('=>' !== $tokenContent) {
+                } elseif (
```
Comment (VincentLanglet) [inline src/Fixer/Operator/BinaryOperatorSpacesFixer.php:862]: The specific behavior won't be hardcoded for `=>` but instead will be an option `BY_SCOPE` for alignment.

File: tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php [hunk 4]
```diff
@@ -1643,6 +1643,703 @@ public function testFixAlignDoubleArrow(string $expected, ?string $input = null)
     }
 
     public static function provideAlignDoubleArrowCases(): array
+    {
+        return [
```
Comment (VincentLanglet) [inline tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php:1647]: The git diff is not great but I've duplicated provideAlignDoubleArrowCases into provideAlignScopedDoubleArrowCases with small diff I'll list.

File: tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php [hunk 5]
```diff
@@ -1643,6 +1643,703 @@ public function testFixAlignDoubleArrow(string $expected, ?string $input = null)
...
+            ],
+            [
+                '<?php
+    return new JsonResponse([
+        "result" => "OK",
+        "html"   => renderView("views/my_view.html.twig", [
+            "foo"    => "bar",
+            "foofoo" => 42,
+        ]),
+        "baz" => "OK",
```
Comment (VincentLanglet) [inline tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php:1714]: In provideAlignScopedDoubleArrowCases, it's aligned with others arrows

File: tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php [hunk 6]
```diff
@@ -1643,6 +1643,703 @@ public function testFixAlignDoubleArrow(string $expected, ?string $input = null)
...
+            ],
+            [
+                '<?php
+    $a = [
+        0              => 1,
+        10 /*Comment*/ => [
+            1  => 2,
+            22 => 3,
+        ],
+        100 => [
```
Comment (VincentLanglet) [inline tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php:1857]: In provideAlignScopedDoubleArrowCases, it's aligned with others arrows

File: tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php [hunk 7]
```diff
@@ -1643,6 +1643,703 @@ public function testFixAlignDoubleArrow(string $expected, ?string $input = null)
...
+    }',
+            ],
+            [
+                '<?php
+    $a = array(
+        10    => 11,
+        20    => 22,
+        30    => 33,
+        40
+            =>
```
Comment (VincentLanglet) [inline tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php:1939]: In provideAlignScopedDoubleArrowCases, it's aligned with others arrows

File: tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php [hunk 8]
```diff
@@ -1643,6 +1643,703 @@ public function testFixAlignDoubleArrow(string $expected, ?string $input = null)
...
+        self::STATUS_UNKNOWN_1 => array("symbol" => "?", "description" => "unknown"),
+        self::STATUS_INVALID_1    => array("symbol" => "III", "description" => "invalid file syntax, file ignored"),
+    );',
+            ],
+            [
+                '<?php
+    $array = array(
+        "bazab" => b(array(
+            1     => 2,
+            5     => [
```
Comment (VincentLanglet) [inline tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php:2009]: In provideAlignScopedDoubleArrowCases, it's aligned with others arrows

**Comment by kenjis** [comment]
Thank you! This is awesome.
I've confirmed our source code style does not break.

> The only doubt I have is if by_scope is a good suffix or if there is one more meaningful...

I'm not good at English, but I'm fine with `by_scope`.

**Comment by VincentLanglet** [comment]
Hi, @julienfalque @keradus 
There is a lot of reaction in the https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6716 issue.
What do you think about this proposal to solve the situation ?

**Comment by VincentLanglet** [comment]
Hi, @julienfalque @keradus 
There was a lot of reaction in the https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6716 issue.
What do you think about this proposal to solve the situation ? 
Is there a way for me to help with this PR ?

**Comment by kenjis** [comment]
We are really waiting for this fix.

**Comment by joaopalopes24** [comment]
We really need this fix. I thought the new organization was not good.

**Comment by keradus** [comment]
TBH this PR is too big complex to find time to properly review it. As you were touching it last, I put my trust on you here. Big thanks for taking care of it, and sorry for delay from our side!

**Comment by VincentLanglet** [comment]
> TBH this PR is too big complex to find time to properly review it. As you were touching it last, I put my trust on you here. Big thanks for taking care of it, and sorry for delay from our side!

Sure, I understand. I tried my best to cover changes with tests. Feel free to ping me if an issue occurs on this rule.

**Pull request merged**